### ISSUE TITLE
Can't make work the Authorize endpoint

### ISSUE THREAD

**Issue by graffiti75** [spotify/web-api#1116]
Issue found on December 20th 2018.

#### Endpoint(s):
* `GET https://accounts.spotify.com/authorize`

#### Scope(s):
* None (application is not using authentication)

#### Steps to reproduce:
1. Open the Postman, a software that processes webservices
2. Type `https://accounts.spotify.com/authorize?client_id=f4ff239b8fbd44a79886b5643c0eeab5&response_type=code&redirect_uri=https%3A%2F%2Fwww.youtube.com`, following [this link](https://developer.spotify.com/documentation/general/guides/authorization-guide/#authorization-code-flow) that explains how to get a TOKEN of the Spotify API.

#### Expected behaviour:
It should be retrieved for me a web service Response with fields:
`code` = An authorization code that can be exchanged for an access token; and 
`state` = The value of the state parameter supplied in the request.

#### Actual behaviour:
```
<!DOCTYPE html>
<html ng-app="accounts" ng-csp>
    <head>
        <meta charset="utf-8">
        <title>Error - Spotify</title>
        <meta name="viewport" content="width=device-width, initial-scale=1.0, maximum-scale=1.0, user-scalable=no">
        <base href="/">
        <link rel="icon" href="https://accounts.scdn.co/images/favicon.ace4d8543bbb017893402a1e9d1ac1fa.ico">
        <link href="https://accounts.scdn.co/css/index.ea78e74e9715b2728580.css" media="screen" rel="stylesheet">
  
    </head>
    <body>
        <div class="head">
            <a class="spotify-logo" href="/" tabindex="-1" title="Spotify"></a>
        </div>
        <div class="container-fluid error">
            <div class="content">
                <h1 class="h1">Error</h1>
                <p>
          Oops! Something went wrong, please try again or check out our 
                    <a href="https://www.spotify.com/help">help area</a>.
                </p>
            </div>
        </div>
        <script async defer src="https://accounts.scdn.co/js/error.b8f16ab4072cd5f96a45.js" sp-error=''></script>
    </body>
</html>
```

**Comment by 0xNF** [comment]
First of all, the redirect URI needs to be a URI that you control in some capacity. I don't know what you're expecting with a redirect that points back to YouTube, unless you intend to intercept it before it gets there somehow.

Second, after attempting to log in, I was greeted with the following OAuth Error:

`
INVALID_CLIENT: Invalid redirect URI
`  

Leading me to suspect that maybe your Redirect URI hasn't been registered with your application. You can check under the dashboard page (in your case, that looks like it would be https://developer.spotify.com/dashboard/applications/f4ff239b8fbd44a79886b5643c0eeab5)  and check under the settings menu to see if your URI is in the list of Redirect URIs.

I don't know if there are other errors, but failing to register the Redirect URI you're using would definitely produce this same error.

**Comment by graffiti75** [comment]
Hello 0xNF.

> First of all, the redirect URI needs to be a URI that you control in some capacity. I don't know what you're expecting with a redirect that points back to YouTube, unless you intend to intercept it before it gets there somehow. 

I will be honest to you: **I am a newbie**.

I am just learning how to use this API. I am learning since yesterday. I am not a Web Developer. I am just an Android Developer learning how to get data from this API. 

My goal here at the moment is to access data from this API into the Web Browser, like Google Chrome.

So, I noticed that to get the TOKEN of access for any search through the API, I need this "redirect url". That's why I used the Youtube url.

Should I use an url like `https://developer.spotify.com/dashboard/applications/f4ff239b8fbd44a79886b5643c0eeab5`? I ask that because that is the API of my Spotify Console.

> Leading me to suspect that maybe your Redirect URI hasn't been registered with your application. You can check under the dashboard page (in your case, that looks like it would be https://developer.spotify.com/dashboard/applications/f4ff239b8fbd44a79886b5643c0eeab5) and check under the settings menu to see if your URI is in the list of Redirect URIs.

I created a project called **ApiTest** in my Spotify Console, as you can see [here](https://user-images.githubusercontent.com/4322963/50301922-ba00b400-046f-11e9-9344-ff9bb1fb74da.png).

Isn't this enough to allow me to get a successful response when trying to get this **Authorize** endpoint?

---

Anyway, I was looking [here](https://developer.spotify.com/documentation/general/guides/authorization-guide/#authorization-code-flow) and I noticed there is 3 ways to get my TOKEN:

**Authorization Flows**
There are 3 optional flows to obtaining app authorization:

- Refreshable user authorization: [Authorization Code](https://developer.spotify.com/documentation/general/guides/authorization-guide/#authorization-code-flow)
- Temporary user authorization: [Implicit Grant](https://developer.spotify.com/documentation/general/guides/authorization-guide/#implicit-grant-flow)
- Refreshable app authorization: [Client Credentials Flow](https://developer.spotify.com/documentation/general/guides/authorization-guide/#client-credentials-flow)

Well, from those 3 ways I think that the third way is the easier one because it just needs 2 parameters to provide my TOKEN:

REQUEST BODY PARAMETER | VALUE
-- | --
`grant_type`	Required. Set it to client_credentials.


HEADER PARAMETER | VALUE
-- | --
`Authorization` Required. Base 64 encoded string that contains the client ID and client secret key. The field must have the format: Authorization: Basic <base64 encoded client_id:client_secret>

Then, I tried this new approach, and **it worked**!

**Comment by 0xNF** [comment]
Congrats on getting to the end of one of the OAuth flows. Navigating OAuth can be nightmarishly hard sometimes, especially to the inexperienced. 

The Client Credentials flow you did is far easier than the Authorization Code flow, but it has some limitations. I'm not sure what your use-case is, but it's possible that you won't run into them and that this is absolutely good enough.

For future reference, you can see what Redirect URIs you have registered under the application's settings section. It looks like this:

<img src="https://i.imgur.com/MEDhjIO.png" height="256"/>

At first it will be empty - you need to add your own. I've already added two localhost based URIs to this one, which you can see in the image.

**Comment by graffiti75** [comment]
My final goal while using this Spotify API is to create an Android App seeking for some musics in Spotify.

At the moment I don't know exactly how I am gonna get that, and how a Redirect URL can help me on that.

**Comment by jscholes** [comment]
> Isn't this enough to allow me to get a successful response when trying to get this Authorize endpoint?

No, because as @0xNF mentioned, the redirect URI is intended to be something which represents a part of your application, e.g. a page on your website or screen in your mobile app which can capture the authorization code and exchange it for the required tokens.  You need to specify your redirect URI in your Spotify Developer Dashboard, and then use the exact same URI when making your request to the authorize endpoint.  If you have no callback URI registered, or you attempt to use one which differs from the one you've registered, the authorization request will fail.

I suppose that, theoretically speaking, you could use youtube.com, as long as you're only intending to capture the authorization code for your own use and then carry out the token exchange manually.  But a better URL for that purpose would be something like:

http://httpbin.org/get

which would show the GET parameters in JSON and not require use of a proxy.

The Client Credentials flow may work fine for your use case, but keep in mind it does not allow any access to private data e.g. a user's library or modification of playlists.

**Comment by 0xNF** [comment]
> http://httpbin.org/get

This is neat.

**Comment by jafioti** [comment]
2 quick questions:
How would the login page even be launched, when my app only receives the raw HTML for the login page?
And once the user logs in, say  I use  https://httpbin.org/get as the redirect URI, would that show the access code in the JSON, and if so how would I be able to retrieve that into my application?

Thanks in advance

**Comment by bih** [comment]
@jafioti If you use the implicit grant flow, you'll receive the access token in the form of an anchor (`#access_token=...` in a URL).

[This is impossible to see in any logs by default](https://webmasters.stackexchange.com/a/696), and this is intentional - an `access_token` is basically just as valuable as a user's password. Other flows require you to do additional steps, and any "exchange or refresh tokens" shared in between are not valuable without your Client ID and Client Secret.

The right way to get it through the implicit grant flow is through JavaScript and calling `window.location.hash` on the URL the user is sent back to.
